src/datumaro/experimental/categories.py

```diff
@@ -131,6 +134,231 @@ def __hash__(self):
         return hash((self.labels, self.group_type, frozenset(self.label_semantics.items())))
 
 
+@dataclass()
+class HierarchicalLabelCategory:
+    """Represents a single label category with hierarchical support."""
+
+    name: str
+    parent: str = field(default="")
+    label_semantics: dict = field(default_factory=dict)
+
+    def __post_init__(self):
+        """Validate that name is not empty."""
+        if not self.name:
+            raise ValueError("Label name cannot be empty")
+
+    def __hash__(self):
+        return hash((self.name, self.parent, frozenset(self.label_semantics.items())))
+
+
+@dataclass()
+class LabelGroup:
+    """Represents a group of labels with a specific group type."""
+
+    name: str
+    labels: Tuple[str, ...] = field(default_factory=tuple)
+    group_type: GroupType = GroupType.EXCLUSIVE
+
+    def __post_init__(self):
+        """Validate that name is not empty and labels is a tuple."""
+        if not self.name:
+            raise ValueError("Label group name cannot be empty")
+        if isinstance(self.labels, list):
+            object.__setattr__(self, "labels", tuple(self.labels))
+        elif not isinstance(self.labels, tuple):
+            raise TypeError("labels must be a tuple of strings")
+
+
+@dataclass()
+class HierarchicalLabelCategories(Categories):
+    """Represents hierarchical label categories with groups and parent-child relationships.
+    Use this for complex hierarchical classification tasks.
+    """
+
+    items: Tuple[HierarchicalLabelCategory, ...] = field(default_factory=tuple)
+    label_groups: Tuple[LabelGroup, ...] = field(default_factory=tuple)
+    label_semantics: dict = field(default_factory=dict)
+
+    def __post_init__(self):
+        """Validate that there are no duplicate labels and build indices."""
+        # Convert lists to tuples if needed
+        if isinstance(self.items, list):
+            object.__setattr__(self, "items", tuple(self.items))
+        if isinstance(self.label_groups, list):
+            object.__setattr__(self, "label_groups", tuple(self.label_groups))
+
+        # Force each group's labels to tuple in case construction happened before class reload
+        for group in self.label_groups:
+            if isinstance(group.labels, list):  # resilience against stale instances
+                object.__setattr__(group, "labels", tuple(group.labels))
+
+        # Validate no duplicate names
+        seen_names = set()
```

Comment:
Is it not better to use label comprehension followed by `set` and comparing the two `len`?

Reply:
Currently, these are not always equal. The old datumaro is sometimes storing label ids as its name, and then the actual name is stored as an attribute. We can use a set once the old datumaro logic has been removed completely.